DAOS-16477 mgmt: return suspect engines for pool healthy query

wangshilong opened the pull request:
After significant failures, the system may leave behind some suspect
engines that were marked as DEAD by the SWIM protocol, but were not
excluded from the system to prevent data loss. An administrator
can bring these ranks back online by restarting them.

This PR aims to provide an administrative interface for querying
suspect engines following a massive failure. These suspect engines
can be retrieved using the daos/dmg --health-only command.

An example of output of dmg pool query --health-only:

Pool 6f450a68-8c7d-4da9-8900-02691650f6a2, ntarget=8, disabled=2, leader=3, version=4, state=Degraded
Pool health info:
- Disabled ranks: 1
- Suspect ranks: 2
- Rebuild busy, 0 objs, 0 recs

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

johannlombardi commented:
@wangshilong reviewing the terminology at the TSC working group and we are suggesting to replace "hanging" in the output of the tool with "suspect".

wangshilong commented:
> @wangshilong reviewing the terminology at the TSC working group and we are suggesting to replace "hanging" in the output of the tool with "suspect".

@johannlombardi Sure

tanabarr reviewed (changes requested):
Changes look okay but we need some updates to documentation to explain what the new parameter means in pool query output and how and why the administrator should use this information.

Also please provide some control-plane unit tests and preferably update ftest for dmg pool query coverage.

mjmac commented on src/control/cmd/daos/pool.go:
```diff
@@ -329,6 +329,9 @@ func queryPool(poolHdl C.daos_handle_t, queryMask daos.PoolQueryMask) (*daos.Poo
 			poolInfo.DisabledRanks = rs
 		}
+		if queryMask.HasOption(daos.PoolQueryOptionSuspectEngines) {
+			poolInfo.SuspectRanks = rs
```
You haven't added the SuspectEngines option to the check up at line 302, so `rlPtr` will always be `nil`.

I know that you're just following the pattern that was here before (which came from a quick refactoring to support pool queries from different parts of the `daos` tool code), but it would be ideal to rework `queryPool()` so that a caller can invoke it once with all of the desired bits set, and then inside of `queryPool()` we can take care of performing multiple pool queries to get the various rank lists. It shouldn't be up to callers of `queryPool()` to do all of this work to build the full response.

Longer-term, I think we need to add a new client-side pool query API that allows all of these rank lists to be fetched at once, similar to how `ds_mgmt_pool_query()` works.

wangshilong replied:
Thanks @mjmac I tried to refactor queryPool(), I agreed with longer-term, let's address that in the separate PR.

liw reviewed (dismissed):
The C changes look good to me.

kjacque reviewed (changes requested):
Agreed with @tanabarr that documentation is needed here to tell users what "suspect" ranks are, and understanding what this means for their system. From reading the code, I get the idea that they are ranks we think may be down, but aren't sure yet for some reason (is this correct?).

I noticed that the suspect ranks were added to the default flags. Do we really want to show these "maybe up, maybe down" ranks by default, if this state is probably transient?

Also, at minimum I think all the Go changes should have matching unit test cases demonstrating the added logic.

wangshilong commented on src/mgmt/tests/srv_drpc_tests.c:
```diff
@@ -1408,12 +1408,13 @@ test_drpc_pool_query_success(void **state)
 	Drpc__Response		resp = DRPC__RESPONSE__INIT;
 	uuid_t			exp_uuid;
 	daos_pool_info_t	exp_info = {0};
+	uint64_t flags = DPI_ENGINES_ENABLED | DPI_ENGINES_DISABLED | DPI_ENGINES_SUSPECT;
```
I tried, clang formatted it back to this.

kjacque replied:
Huh, okay. It's weird how clang-format seems to defy its own rules sometimes.

wangshilong commented on src/pool/srv_cli.c:
```diff
@@ -367,16 +368,65 @@ pool_query_init(uuid_t pool_uuid, crt_rpc_t *rpc, void *varg)
+
+		D_DEBUG(DB_MD, "rank/state %d/%d\n", doms[i].do_comp.co_rank,
+			rc == -DER_NONEXIST ? -1 : state.sms_status);
+		if (rc == -DER_NONEXIST || state.sms_status == SWIM_MEMBER_DEAD) {
```
To be honest, i don't quite understand this too, i copied this logic from Xuezhao's codes, @liuxuezhao would you help answer this question please?

liw replied:
For example, the pool comprises rank 0, 1, and 2. Rank 2, however, has been excluded from the system by the MS, and has therefore been deleted from the current engine's local primary group. In this case, we'll get `-DER_NONEXIST` here. In other words, both `-DER_NONEXIST` and "exist but dead" imply that the rank is "unavailable".

kjacque replied:
I see. Thanks for the clarification, @liw!

wangshilong commented:
> Agreed with @tanabarr that documentation is needed here to tell users what "suspect" ranks are, and understanding what this means for their system. From reading the code, I get the idea that they are ranks we think may be down, but aren't sure yet for some reason (is this correct?).

I added documents in the following commits.

> 
> I noticed that the suspect ranks were added to the default flags. Do we really want to show these "maybe up, maybe down" ranks by default, if this state is probably transient?

Yes, it is transient. Notice this is only added to health query default flags.

> 
> Also, at minimum I think all the Go changes should have matching unit test cases demonstrating the added logic.

Done

wangshilong commented:
> Changes look okay but we need some updates to documentation to explain what the new parameter means in pool query output and how and why the administrator should use this information.

I added example and documents in the "pool operation"

> 
> Also please provide some control-plane unit tests and preferably update ftest for dmg pool query coverage.

Added, control-plane unit tests, I am trying to update ftests which might cause more time(I might test commits in the PR or in a separate PR) Thanks.

tanabarr commented on src/control/cmd/daos/pool.go:
```diff
@@ -337,6 +338,55 @@ func queryPool(poolHdl C.daos_handle_t, queryMask daos.PoolQueryMask) (*daos.Poo
 	return poolInfo, nil
 }
 
+func queryPool(poolHdl C.daos_handle_t, queryMask daos.PoolQueryMask) (*daos.PoolInfo, error) {
```
do these really need to be done sequentially, can't the mask handle multiple options set in the same query?

wangshilong replied:
We could do that in the long-term goal, introduce daos_pool_query_v2() to return three rank lists at the same time.

tanabarr reviewed (changes requested):
changes look good just some minor updates required for doc updates

liw commented on docs/admin/pool_operations.md:
```diff
@@ -286,6 +286,28 @@ The example below shows a rebuild in progress and NVMe space allocated.
 
+After experiencing significant failures, the system may retain some suspect
+engines that have been marked as DEAD by the SWIM protocol but were not excluded
+from the system to prevent data loss. An administrator can bring these engines back
```
Agreed. Plus, two nit-picking comments below, just for consideration.

"The system may retain" and "from the system" are technically incorrect. Such an engine is not excluded **from the pool**, but may have been excluded from the system. So "the pool" might be more accurate.

"To prevent data loss" could be more accurate as "to prevent potential data inconsistency" or "to preserve data consistency". The foremost purpose of not excluding above RF is to prevent applications from reading potential data that would be impacted if we were to exclude all unavailable engines. The second purpose is to avoid having to reintegrate too many engines to work around certain aspects of the current reintegration protocol.

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15196/7/execution/node/1173/log

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15196/8/execution/node/1202/log

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15196/9/testReport/

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15196/9/execution/node/1197/log

tanabarr commented on src/tests/ftest/control/dmg_pool_query_ranks.py:
```diff
@@ -80,58 +87,77 @@ def test_pool_query_ranks_mgmt(self):
+
+        # kill second rank.
+        self.server_managers[0].stop_ranks([suspect_rank], self.d_log)
+        time.sleep(20)
```
Isn't there a better way than using fixed sleeps?? seems like we might be introducing intermittent failures by using them

daltonbohning replied:
I agree with Tom. It would be better if we can deterministically decide when the rank has stopped. It's possible to check the corresponding log file, but I would need to dig up the relevant functions. Or do we think 20 seconds is almost always enough, even for various configs?

wangshilong replied:
Added a new helper for that.

tanabarr commented on src/tests/ftest/control/dmg_pool_query_ranks.py:
```diff
@@ -80,58 +87,77 @@ def test_pool_query_ranks_mgmt(self):
+                disabled_ranks, data['response'].get('disabled_ranks')))
+
+        self.server_managers[0].start_ranks([suspect_rank], self.d_log)
+        time.sleep(10)
```
as above

daltonbohning replied:
I think this is why the previous code used
```
self.pool.wait_for_rebuild_to_start()
self.pool.wait_for_rebuild_to_end()
```

wangshilong replied:
..

tanabarr reviewed (approved):
nice work!

daltonbohning commented on src/tests/ftest/control/dmg_pool_query_ranks.py:
```diff
@@ -23,7 +23,7 @@ def setUp(self):
 
         # Init the pool
-        self.add_pool(connect=False)
+        self.add_pool(connect=False, properties='rd_fac:1')
```
We should instead put the properties in the config `dmg_pool_query_ranks.yaml`. Under the existing `pool` key:
```
pool:
  properties: rd_fac:1
```

wangshilong replied:
Fixed.

daltonbohning commented on src/tests/ftest/control/dmg_pool_query_ranks.py:
```diff
@@ -80,58 +87,77 @@ def test_pool_query_ranks_mgmt(self):
+            "Invalid disabled_ranks field: want={}, got={}".format(
+                disabled_ranks, data['response'].get('disabled_ranks')))
+
+        self.log.debug("Waiting for pool to be rebuild")
+        self.pool.wait_for_rebuild_to_start()
```
Only wait for rebuild to start? And don't wait for it to finish?

wangshilong replied:
Yes, that is written on purpose to see suspect ranks.

daltonbohning commented on src/tests/ftest/control/dmg_pool_query_ranks.py:
```diff
@@ -80,58 +87,77 @@ def test_pool_query_ranks_mgmt(self):
+                self.log.debug("dmg command failed retry")
+            time.sleep(3)
+
+        self.assertTrue(cmd_succeed, "pool reintegrate failed: {}".format(result))
```
If `self.pool.reintegrate(exclude_rank)` failed above and raised an exception, `result` will not be set. I think in this case we don't need to log it, since the output will be in the test log
```suggestion
        self.assertTrue(cmd_succeed, "pool reintegrate failed")
```

wangshilong replied:
Fixed.

daltonbohning commented on src/tests/ftest/control/dmg_pool_query_ranks.yaml:
```diff
@@ -1,20 +1,26 @@
   system_ram_reserved: 1
 pool:
   size: 4GB
+  svcn: 5
```
I think better to not set svcn, since the default will be 5 with 5 servers? This has been an issue before, so we've started removing svcn from pool create and letting DAOS handle it

wangshilong replied:
This is written on purpose, i want to simplify codes, otherwise i need take care in the codes don't break 1/2 of PS replica, other pool query won't work.

daltonbohning commented on src/tests/ftest/util/dmg_utils_base.py:
```diff
@@ -6,8 +6,10 @@
 from socket import gethostname
 
 from ClusterShell.NodeSet import NodeSet
+
 from command_utils import CommandWithSubCommand, YamlCommand
-from command_utils_base import BasicParameter, CommandWithParameters, FormattedParameter
+from command_utils_base import (BasicParameter, CommandWithParameters,
+                                FormattedParameter)
 from environment_utils import TestEnvironment
```
> Imports are incorrectly sorted and/or formatted.

I think it wants this
```suggestion
from ClusterShell.NodeSet import NodeSet
from command_utils import CommandWithSubCommand, YamlCommand
from command_utils_base import BasicParameter, CommandWithParameters, FormattedParameter
```

wangshilong replied:
Sure.

daltonbohning commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -5,7 +5,8 @@
 import os
 
-from command_utils_base import BasicParameter, LogParameter, TransportCredentials, YamlParameters
+from command_utils_base import (BasicParameter, LogParameter,
+                                TransportCredentials, YamlParameters)
```
```suggestion
from command_utils_base import BasicParameter, LogParameter, TransportCredentials, YamlParameters
```

daltonbohning replied:
You might have a different version of isort if the githooks are changing this.

You can check with
```
python3 -m isort --version
```

And the current you should have is 5.13.2, from `utils/cq/requirements.txt`

daltonbohning reviewed (changes requested):
ftest changes LGTM but since this touches more that just ftest, we should use `Features` instead of `Test-tag`
```
Features: DmgPoolQueryRanks
```
`Test-tag` tells CI to run ONLY that one test. Also, this test is automatically included because you modified it so you don't even need `Features`

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15196/16/execution/node/1351/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15196/16/execution/node/1397/log

wangshilong commented on src/tests/ftest/util/server_utils_params.py:
```diff
@@ -495,6 +494,7 @@ def __init__(self, base_namespace, index, provider=None, max_storage_tiers=MAX_S
             "ABT_ENV_MAX_NUM_XSTREAMS=100",
             "ABT_MAX_NUM_XSTREAMS=100",
             "DAOS_MD_CAP=1024",
+            "DAOS_POOL_RF=4",
```
@daltonbohning would you help point me how i could modify this parts? I notice if i moved DAOS_POOL_RF=4 to here, some tests, eg rebuild won't inherit this proper：
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15196/16/artifact/Functional%20Hardware%20Medium%20Verbs%20Provider/daos_test/suite.py/daos_configs.wolf-217/daos_server.yml/*view*/

But i want to specify this in dmg pool query test, all other tests should be DAOS_POOL_RF=4, it looks a messy to me.

daltonbohning replied:
I think the problem is this is the default set of `env_vars`, but since https://github.com/daos-stack/daos/blob/master/src/tests/ftest/daos_test/suite.yaml#L51-L57 defines `env_vars` these defaults aren't used.

@phender Is this the intended behavior?

liuxuezhao replied:
Just curious why CRT_EVENT_DELAY and DAOS_POOL_RF set at different places?

kjacque commented on src/control/cmd/daos/pool.go:
```diff
@@ -329,6 +330,58 @@ func queryPool(poolHdl C.daos_handle_t, queryMask daos.PoolQueryMask) (*daos.Poo
+	origQueryMask := queryMask
+	queryMask.ClearOptions(daos.PoolQueryOptionEnabledEngines,
+		daos.PoolQueryOptionSuspectEngines, daos.PoolQueryOptionDisabledEngines)
+	if origQueryMask.HasOption(daos.PoolQueryOptionEnabledEngines) {
+		queryMask.SetOptions(daos.PoolQueryOptionEnabledEngines)
+		origQueryMask.ClearOptions(daos.PoolQueryOptionEnabledEngines)
+	} else if origQueryMask.HasOption(daos.PoolQueryOptionDisabledEngines) {
+		queryMask.SetOptions(daos.PoolQueryOptionDisabledEngines)
+		origQueryMask.ClearOptions(daos.PoolQueryOptionDisabledEngines)
+	} else if origQueryMask.HasOption(daos.PoolQueryOptionSuspectEngines) {
+		queryMask.SetOptions(daos.PoolQueryOptionSuspectEngines)
+		origQueryMask.ClearOptions(daos.PoolQueryOptionSuspectEngines)
+	}
+	poolInfo, err := queryPoolHelper(poolHdl, queryMask)
+	if err != nil {
+		return nil, err
+	}
```
I don't really understand why this part is necessary given that we go through the sequence of all three flags below.

wangshilong replied:
It is possible that we don't have all three flags enabled at all.. Notice if we only have one flag enabled, after we run this query, the following query won't be executed since we cleared original flag.

kjacque replied:
Ah, I see. Thanks for the explanation. I think refactoring this area to reduce repetition will probably help make things clearer.

kjacque commented on src/control/cmd/daos/pool.go:
```diff
@@ -329,6 +330,58 @@ func queryPool(poolHdl C.daos_handle_t, queryMask daos.PoolQueryMask) (*daos.Poo
+	if err != nil {
+		return nil, err
+	}
+	if origQueryMask.HasOption(daos.PoolQueryOptionEnabledEngines) {
+		queryMask.ClearAll()
+		queryMask.SetOptions(daos.PoolQueryOptionEnabledEngines)
+		poolInfo1, err := queryPoolHelper(poolHdl, queryMask)
+		if err != nil {
+			return nil, err
+		}
+		poolInfo.EnabledRanks = poolInfo1.EnabledRanks
+	}
+	if origQueryMask.HasOption(daos.PoolQueryOptionDisabledEngines) {
+		queryMask.ClearAll()
+		queryMask.SetOptions(daos.PoolQueryOptionDisabledEngines)
+		poolInfo1, err := queryPoolHelper(poolHdl, queryMask)
+		if err != nil {
+			return nil, err
+		}
+		poolInfo.DisabledRanks = poolInfo1.DisabledRanks
+	}
+	if origQueryMask.HasOption(daos.PoolQueryOptionSuspectEngines) {
+		queryMask.ClearAll()
+		queryMask.SetOptions(daos.PoolQueryOptionSuspectEngines)
+		poolInfo1, err := queryPoolHelper(poolHdl, queryMask)
+		if err != nil {
+			return nil, err
+		}
+		poolInfo.SuspectRanks = poolInfo1.SuspectRanks
```
This seems like it could be a loop or a helper function, as these three blocks are very similar.

wangshilong replied:
Sure.

kjacque commented on src/control/cmd/daos/pool.go:
```diff
@@ -295,11 +295,12 @@ func convertPoolInfo(pinfo *C.daos_pool_info_t) (*daos.PoolInfo, error) {
 }
 
-func queryPool(poolHdl C.daos_handle_t, queryMask daos.PoolQueryMask) (*daos.PoolInfo, error) {
+func queryPoolHelper(poolHdl C.daos_handle_t, queryMask daos.PoolQueryMask) (*daos.PoolInfo, error) {
```
naming suggestion - I'd leave `queryPool` as-is, as this is the function that calls `daos_pool_query`. Maybe the other function could be called `queryPoolRankLists`, since its main goal is to collect the different rank lists from multiple queries.

wangshilong replied:
OK

kjacque reviewed (commented):
Thanks for the fixes. I have a couple more comments on the repetition in the pool query logic. Would be nice to clean this up a bit before landing.

wangshilong commented:
ping reviewers..

tanabarr commented on src/control/cmd/daos/pool.go:
```diff
@@ -337,51 +337,61 @@ func queryPoolHelper(poolHdl C.daos_handle_t, queryMask daos.PoolQueryMask) (*da
+	poolInfo := &daos.PoolInfo{}
+	originalMask := queryMask // Save the original queryMask
+
+	// Function to handle the query and return a single RankList
```
nit: In order to avoid unintended consequences, in-line functions like this should accept the object they are acting on as a parameter rather than mutating objects in an outer scope.

kjacque replied:
This is a good point. I think in this case it is safe, but it is easier to reason about if it accepts the object to be acted upon.

liw reviewed (approved):
The C changes look good to me.

kjacque reviewed (approved):
OK with me. Thanks.
